src/Core/SymbolUtility.cs

```diff
@@ -623,53 +623,57 @@ public static ulong GetEnumValueAsUInt64(object value, INamedTypeSymbol enumType
         }
     }
 
-    public static bool IsAwaitable(ITypeSymbol typeSymbol, bool shouldCheckWindowsRuntimeTypes = false)
+    /// <summary>
+    /// Determines if the symbol is an awaitable type.<br/>
+    /// Any type can be <see langword="await"/>ed if it has an instance or extension <c>GetAwaiter</c> method that returns a correctly-shaped awaiter type.
+    /// </summary>
+    /// <remarks>
+    /// For more information, see the <see href="https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#12982-awaitable-expressions">C# language specification</see>.
+    /// </remarks>
+    /// <param name="semanticModel">Used to determine whether a candidate <c>GetAwaiter</c> method is accessible at the given <paramref name="position"/>.</param>
+    /// <param name="position">The position of the token at which the <paramref name="symbol"/> is used.</param>
+    /// <returns>A <see cref="bool"/> indicating whether the symbol is an awaitable type or a method that returns one.</returns>
+    public static bool IsAwaitable(this ISymbol? symbol, SemanticModel semanticModel, int position)
     {
-        INamedTypeSymbol? namedTypeSymbol = GetPossiblyAwaitableType(typeSymbol);
+        ITypeSymbol? typeSymbol = (symbol as ITypeSymbol) ?? (symbol as IMethodSymbol)?.ReturnType;
 
-        if (namedTypeSymbol is null)
+        if (typeSymbol is null or { SpecialType: SpecialType.System_Void })
```

Comment:
`GetAwaiter` can be an extension method (as is the case with WinRT), so something like this would actually work:
```csharp
public static TaskAwaiter GetAwaiter(this int millis)
{
    return Task.Delay(millis).GetAwaiter();
}

public static async Task Example()
{
    await 2000;
}
```
Not going to vouch for that as particularly great code... but it's valid and other variations can exist.

Reply:
Interesting.